### ISSUE TITLE
google_checks: update to most recent version of style guide (3 November 2016)

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3755]
http://checkstyle.sourceforge.net/google_style.html
https://github.com/google/styleguide/commits/gh-pages/javaguide.html

change that we miss: https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c

**Comment by rnveach** [comment]
Our copy of the guide was 2016-07-12 which corresponds to https://github.com/google/styleguide/commit/b075cb7cad464b224103fb2ec8fdb8521ec08e7f .

To see the differences between that version and current master, you can use the following URL:
https://github.com/google/styleguide/compare/b075cb7cad464b224103fb2ec8fdb8521ec08e7f...gh-pages#diff-b6c9191ee5ccb7a5ef170e98e51d94d8

**Comment by MEZk** [comment]
I'm on it.

**Comment by rnveach** [comment]
@MEZk Yes, both are separate. This issue should apply changes up to date specified in title.
#3888 will do from this date to the next one in that title.

> What will we do if changes in #3888 will be completely controversial to changes in this issue?

IMO, let us know if there are any. I haven't looked at both to see what is involved.

**Comment by MEZk** [comment]
@rnveach @romani 
I reviewed all changes. Here is a quick summary:


Changes in Google Java Syle Guide which do not require additional changes in google_checks.xml or xdoc:

1) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R40
2) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R44
3) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R225
4) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R229
5) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R240
6) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R394
7) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R460
8) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R473
9) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R671
10) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R847
11) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R1105
12) https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R671


Changes in Google Java Syle Guide which require changes in xdoc and/or the implementation of a new logic:

1) [3.3.4 No static import for classes](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R210) requires an implementation of a new option for AvoidStaticImportCheck. The main idea is to check whether there is an import of a class or there is an import of other static member. Now we check [both](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/AvoidStaticImportCheck.java#L123)
2) [4.1.3 Empty blocks: may be concise](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R312) requires [new exmaples](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R312) in IT area.
3) [This change](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R422) will require the implementation of the logic which will process lambdas for [NoLineWrapCheck](http://checkstyle.sourceforge.net/config_whitespace.html#NoLineWrap) or [OperatorWrap](http://checkstyle.sourceforge.net/config_whitespace.html#OperatorWrap). Now they do not treat lambdas.

The following change in Google Java Style Guide is not mandatory:
1) [This](https://github.com/google/styleguide/commit/db0a26320f3e930c6ea7225ed53539b4fb31310c#diff-b6c9191ee5ccb7a5ef170e98e51d94d8R744) says that any line break may (but not must) be preceded by arbitrary whitespace followed by an implementation comment. So it is not mandatory.

Few more questions:
1) How can we update the cached version of Google Java Style Guide stored at http://checkstyle.sourceforge.net/reports/google-java-style-20160712.html
2) Should we create new issues for the rules which we cannot use now and link them as blockers at http://checkstyle.sourceforge.net/google_style.html

**Comment by romani** [comment]
> requires an implementation of a new option for AvoidStaticImportCheck. The main idea is to check whether there is an import of a class or there is an import of other static member. 

I am ok with new option. You can split it out in separate issue. 

> This change will require the implementation of the logic which will process lambdas for NoLineWrapCheck or OperatorWrap. Now they do not treat lambdas.

separate to 2 issues.

>  says that any line break may (but not must) be preceded by arbitrary whitespace followed by an implementation comment. So it is not mandatory.

we need to ease our config.

> How can we update the cached version of Google Java Style Guide stored at 

I do this manually, as I receive PR that covers new version of document. As we can not cover all in one issue, it is ok to update coverage page with links to issue where it will be covered

> Should we create new issues for the rules which we cannot use now and link them as blockers

not as blockers, but as future plans.

==========================

in this issue , I suggest to update coverage with links to issues when coverage become complete and I do cache of guide.

**Comment by rnveach** [comment]
@romani 
> > How can we update the cached version of Google Java Style Guide stored at

> I do this manually

Is there any reason we can't put this as a pure `html` file in git repository, like under the `site` directory?
We wouldn't need a date linked file as we could see the changes directly in git.

**Comment by romani** [comment]
> Is there any reason we can't put this as a pure html file in git repository, like under the site directory?

no good reason, it was required to be done ones manually (3 years ago), after that I updated it only ones. Now looks like Google team is restored interest to guide, so we could make it more automated. Please make it as separate issue.

**Comment by MEZk** [comment]
@romani 
> I am ok with new option. You can split it out in separate issue.

Done. #4005
